### ISSUE TITLE
Incorrect reportOptionalSubscript when accessing List with checked length.

### ISSUE THREAD

**Issue by NielsKorschinsky** [microsoft/pylance-release#3077]
<h3> Environment data </h3>

-   Language Server version: 2022.7.30
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: null
-   python.analysis.typeCheckingMode: strict

<h3> Code Snippet </h3>

```python

group_tags: Optional[List[str]] = ["hello", "\"anotherVal\""]
tag_dict: Optional[Dict[str, str]] = {"hello": "wantedId", "anotherVal": "wantedName"}
insert_tags: Dict[str, Optional[str]] = {}

if group_tags and tag_dict:
    if "something" in group_tags:
        pass
    else:
        insert_tags["id_value"] = tag_dict[group_tags[0].replace("\"", "")]
        if len(group_tags) > 1:
            insert_tags["name_value"] = tag_dict[group_tags[1].replace("\"", "")]
```

Error on the part `group_tags` (last line, accessing [1])
```
(parameter) group_tags: List[str]
Object of type "None" is not subscriptablePylance[reportOptionalSubscript](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)
``` 

<h3> Logs </h3>

I will add it if requested.

```
This setting cannot be applied in this workspace. It will be applied when you open the containing workspace folder directly.
```

**Comment by NielsKorschinsky** [comment]
Deleting the `replace` removes this bug, though the marker is only on the `group_tags`-part.

**Comment by NielsKorschinsky** [comment]
Ah, I had an error in the first variable declaration. I have fixed it now, please try again to reproduce

**Comment by erictraut** [comment]
To confirm, you're able to see the error with the self-contained minimal code sample above? I don't see any error with this code. Please try pasting this code into its own ".py" file and make sure you can repro the problem there. If you can, we'll need to figure out what's different between your configuration and mine.

**Comment by NielsKorschinsky** [comment]
I tried and I am not able to reproduce either. 
I need to add a tuple-unpacking in a loop to reproduce it.

I cannot attach the file, therefore here is the full code 1-1:

```python

from typing import Optional, List, Dict, Tuple, Any

group_tags: Optional[List[str]] = ["hello", "\"anotherVal\""]
insert_tags: Dict[str, Optional[str]] = {}

tag_data_tuple: List[Tuple[
                    Optional[Dict[str, str]],
                    List[Dict[str, Any]]]] = [
    ({"hello": "wantedId", "anotherVal": "wantedName"}, [{}])
]

for (tag_dict, data) in tag_data_tuple:

    if group_tags and tag_dict:
        if "something" in group_tags:
            pass
        else:
            insert_tags["id_value"] = tag_dict[group_tags[0].replace("\"", "")]
            if len(group_tags) > 1:
                insert_tags["name_value"] = tag_dict[group_tags[1].replace("\"", "")]

```

**Comment by erictraut** [comment]
OK, thanks. I'm able to repro the problem with that code sample.

This is a core type checker issue, so I've created a [tracking bug](https://github.com/microsoft/pyright/issues/3731) in the pyright repo.

**Comment by NielsKorschinsky** [comment]
Thanks - I am surprised, I though that pyright was integrated into pylance.

**Comment by erictraut** [comment]
Yes, pyright is integrated into pylance. Or perhaps it would be more accurate to say that pylance is built on top of pyright, at least from an architectural layering perspective. All of the core type analysis logic is in pyright, and pylance builds features on top of that capability. This particular bug is in the pyright logic.

**Comment by erictraut** [comment]
This bug will be fixed in the next release of pyright and will be picked up by pyright in some future release.

**Comment by heejaechang** [comment]
This issue has been fixed in version 2022.7.42, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2022742-27-July-2022)
